internal/util/update/update_test.go

```diff
@@ -2009,3 +2009,249 @@ func TestReplaceNonKRMFiles(t *testing.T) {
 		})
 	}
 }
+
+func TestCommand_Run_updateSubpackage(t *testing.T) {
+	testCases := map[string]struct {
+		initialUpstream *pkgbuilder.RootPkg
+		refRepos        map[string][]testutil.Content
+		subPkgPath      string
+		repoRef         string
+		ref             string
+		expectedResult  *pkgbuilder.RootPkg
+		expectedErrMsg  string
+	}{
+		"updating a subpackage should update parent Kptfile": {
+			initialUpstream: pkgbuilder.NewRootPkg().
+				WithKptfile(
+					pkgbuilder.NewKptfile().
+						WithSubpackages(
+							pkgbuilder.NewSubpackage("foo", "/", "master", "resource-merge", "foo"),
+						),
+				).
+				WithResource(pkgbuilder.DeploymentResource),
+			refRepos: map[string][]testutil.Content{
+				"foo": {
+					{
+						Pkg: pkgbuilder.NewRootPkg().
+							WithResource(pkgbuilder.ConfigMapResource),
+						Branch: "master",
+					},
+					{
+						Pkg: pkgbuilder.NewRootPkg().
+							WithResource(pkgbuilder.DeploymentResource),
+						Tag: "my-tag",
+					},
+				},
+			},
+			subPkgPath: "foo",
+			repoRef:    "foo",
+			ref:        "my-tag",
+			expectedResult: pkgbuilder.NewRootPkg().
+				WithKptfile(
+					pkgbuilder.NewKptfile().
+						WithUpstreamRef("upstream", "/", "master").
+						WithSubpackages(
+							pkgbuilder.NewSubpackage("foo", "/", "my-tag", "resource-merge", "foo"),
+						),
+				).
+				WithResource(pkgbuilder.DeploymentResource).
+				WithSubPackages(
+					pkgbuilder.NewSubPkg("foo").
+						WithKptfile(
+							pkgbuilder.NewKptfile().
+								WithUpstreamRef("foo", "/", "my-tag"),
+						).
+						WithResource(pkgbuilder.DeploymentResource),
+				),
+		},
+		"changes to Kptfile are rolled back if update is not successful": {
```

Comment:
Does this mean, upon calling update on child package, 
1. Parent package Kptfile is updated
2. Subpackage is updated
3. If step 2 fails, Step 1 is reverted?

Why not just update subpackage, and then update the parent package Kptfile as updating parent Kptfile is less likely to fail.

Reply:
Yeah, for now it seems like this is a bit cumbersome, but it follows the model I think we want to use. So most commands should follow two steps (like setters does today), where the first is to update the declarative state and the second is to make it happen. Ideally we should let the update `Command` take in a `Subpackage` as defined in the parent Kptfile, but that means we still need a way to handle the root package.

Longer term I think we should consider handling errors by doing the full update process in a temp location and then just move things over when it completed.

Reply:
Makes sense.